### PR TITLE
support if logic for Variable in dygraph

### PR DESCRIPTION
**Pull request by JepsonWong**
support if logic for Variable in dygraph.

```
with fluid.dygraph.guard():
    var1 = fluid.dygraph.to_variable(np.array([[[0]]]))
    var1_bool = False
    if var1:
        var1_bool = True
```

related pr: https://github.com/PaddlePaddle/models/pull/4447

### CODE REVIEW

File: python/paddle/fluid/dygraph/varbase_patch_methods.py [hunk 1]
```diff
@@ -267,10 +267,21 @@ def __getitem__(self, item):
 
         return out
 
-    for method_name, method in (("set_value", set_value), ("block", block),
-                                ("backward", backward), ("gradient", gradient),
-                                ("__str__", __str__), ("to_string", to_string),
-                                ("__getitem__", __getitem__)):
+    def __nonzero__(self):
+        numel = np.prod(self.shape)
+        assert numel == 1, "If you want to use if logic to determine whether the variable is None type, please use code like if variable is None or if Variable is not None. If you want to use if logic for Variable, Variable can only contain one element."
```
Comment (phlrain) [inline python/paddle/fluid/dygraph/varbase_patch_methods.py]: 这个不仅仅是if会用到， 也可能是作为 while的条件

Comment (JepsonWong) [reply python/paddle/fluid/dygraph/varbase_patch_methods.py]: done

File: python/paddle/fluid/dygraph/varbase_patch_methods.py [hunk 2]
```diff
@@ -204,9 +204,20 @@ def to_string(self, throw_on_error, with_details=False):
                 return 'name %s, shape: %s, not inited' % (self.name,
                                                            self.shape)
 
-    for method_name, method in (("set_value", set_value), ("block", block),
-                                ("backward", backward), ("gradient", gradient),
-                                ("__str__", __str__), ("to_string", to_string)):
+    def __nonzero__(self):
+        numel = np.prod(self.shape)
+        assert numel == 1, "If you want to use if/while logic to determine whether the variable is None type, please use code like if variable is None or if Variable is not None. If you want to use if/while logic for Variable, Variable can only contain one element."
```
Comment (phlrain) [inline python/paddle/fluid/dygraph/varbase_patch_methods.py]: "When Variable is used as the condition of if/while , Variable can only contain one element."
error  message 换成这个吧

Comment (JepsonWong) [reply python/paddle/fluid/dygraph/varbase_patch_methods.py]: done

**Review by phlrain — DISMISSED** [review]

**Review by phlrain — APPROVED** [review]

**Review by phlrain — APPROVED** [review]

**Pull request merged**